src/main/java/teammates/logic/core/AccountsLogic.java

```diff
@@ -238,4 +242,49 @@ private void createStudentAccount(StudentAttributes student)
         accountsDb.createEntity(account);
     }
 
+    /**
+     * Updates the readNotifications of an account.
+     *
+     * @param googleId google ID of the user who read the notification.
+     * @param notificationId notification to be marked as read.
+     * @param endTime the expiry time of the notification, i.e. notification will not be shown after this time.
+     * @return the account attributes with updated read notifications.
+     * @throws InvalidParametersException if the notification has expired.
+     * @throws EntityDoesNotExistException if account or notification does not exist.
+     */
+    public AccountAttributes updateReadNotifications(String googleId, String notificationId, Instant endTime)
+            throws InvalidParametersException, EntityDoesNotExistException {
+        AccountAttributes a = accountsDb.getAccount(googleId);
+
+        if (a == null) {
+            throw new EntityDoesNotExistException("Trying to update the read notifications of a non-existent account.");
+        }
+        if (!notificationsLogic.doesNotificationExists(notificationId)) {
+            throw new EntityDoesNotExistException("Trying to mark as read a notification that does not exist.");
+        }
+        if (endTime.isBefore(Instant.now())) {
+            throw new InvalidParametersException("Trying to mark an expired notification as read.");
+        }
+
+        Map<String, Instant> updatedReadNotifications = a.getReadNotifications();
+        // expired notifications are removed from readNotifications as inactive notification will no longer be fetched.
+        for (Map.Entry<String, Instant> notification : updatedReadNotifications.entrySet()) {
+            if (notification.getValue().isBefore(Instant.now())) {
+                updatedReadNotifications.remove(notification.getKey());
+            }
+        }
+
+        updatedReadNotifications.put(notificationId, endTime);
+
+        try {
+            return accountsDb.updateAccount(
+                    AccountAttributes.updateOptionsBuilder(googleId)
+                            .withReadNotifications(updatedReadNotifications)
+                            .build());
+        } catch (InvalidParametersException | EntityDoesNotExistException e) {
+            assert false : "Unexpected error "
+                    + "while updating read notifications for account " + googleId + ": " + e.getMessage();
+        }
+        throw new EntityDoesNotExistException("Invalid update to read notification in account");
+    }
```

Comment:
It feels weird and redundant to throw `EntityDoesNotExistException` again when it was already caught in the catch block above, perhaps you can do something like this:

```suggestion
        AccountAttributes newAccountAttributes = a.getCopy();

        try {
            newAccountAttributes = accountsDb.updateAccount(
                    AccountAttributes.updateOptionsBuilder(googleId)
                            .withReadNotifications(updatedReadNotifications)
                            .build());
        } catch (InvalidParametersException | EntityDoesNotExistException e) {
            assert false : "Unexpected error "
                    + "while updating read notifications for account " + googleId + ": " + e.getMessage();
        }
        
        return newAccountAttributes;
    }
```

Reply:
In that case, since the method signature includes `InvalidParametersException, EntityDoesNotExistException`, would it be cleaner to just return the result without try-catch blocks?
```suggestion
        return accountsDb.updateAccount(
                AccountAttributes.updateOptionsBuilder(googleId)
                        .withReadNotifications(updatedReadNotifications)
                        .build());
```